client/registry.go

```diff
@@ -0,0 +1,243 @@
+package client
+
+import (
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/mysteriumnetwork/payments/bindings"
+)
+
+type HermesImplementationRegistry struct {
+	callers     map[string]*bindings.HermesImplementationCaller
+	filterers   map[string]*bindings.HermesImplementationFilterer
+	transactors map[string]*bindings.HermesImplementationTransactor
+}
+
+func NewHermesImplementationRegistry() *HermesImplementationRegistry {
+	return &HermesImplementationRegistry{
+		callers:     make(map[string]*bindings.HermesImplementationCaller),
+		filterers:   make(map[string]*bindings.HermesImplementationFilterer),
+		transactors: make(map[string]*bindings.HermesImplementationTransactor),
+	}
+}
+
+func (h *HermesImplementationRegistry) Caller(address common.Address, etherClient EtherClient) (*bindings.HermesImplementationCaller, error) {
+	if caller, exists := h.callers[address.Hex()]; exists {
```

Comment:
on first look it seems like calling this function with eth client and then with polygon client would make it return a wrong client no?

Reply:
We can say so, but do we have same address in two chains somewhere? What is the possibility of it?

Reply:
we do, check [here](https://github.com/mysteriumnetwork/payments-smart-contracts). For example registry is the same in both chains

Reply:
also another problem might be if a node is down and we start using another rpc, for the same chain. I'm not sure how to solve this one...

Reply:
Good 👍 , will consider this

Reply:
We use, eth.ethswitch.svc for different RPC. Does this solve it?

Reply:
in our case yes, but seems like bad practice to code based on our particular approach, maybe it would be nice to have the client rpc URL in the key somehow

Reply:
EtherClient contains multiple rpc addreses

Reply:
If we use direct RPC it will fail (in acse of rpc down) even now without this changes. If we use ethswitch - this solves it

Reply:
maybe each etherclient could have an id or something to identify them. Or else as you say we can assume 1 client per chain I guess

Reply:
But how to change this ID? We are not constructing or changing BC Client (RPC) dynamically. It never changes during app lifecycle (or it is somewhere) ?

Reply:
just set it on creation, anyway I'm not sure if its a good idea, we can discuss it tomorrow and see what is best. I will try to take a look at the code a bit deeper too

Reply:
`EtherClient` is immutable. How do you think we can dynamically change it?

Reply:
I am not suggesting to change it, just make each client identifiable so it can be a part of the cache key

Reply:
on creation of the etherclient I meant

Reply:
I agree to include chain id, that's good.
But for RPC: to make your case real we need to implement somewhere logic for changing ethclient on some condition (fail or something). Do we have something? Do we need implement it if we have ethswith for this?